### PR TITLE
[SG2] Add non-GraphType blackboard properties to menu

### PR DESCRIPTION
**Pull request by dhsavell**
### Purpose of this PR

This PR adds blackboard support for a handful of non-GraphType properties. Compared to SG1, the only types still missing are Diffusion Profile and Virtual Texture.

This PR doesn't make any UI changes yet. So the "exposed" buttons do nothing, some editors are still incorrect, and the new types don't expose any of their associated settings.

Note: some types can break the asset similarly to https://jira.unity3d.com/browse/GSG-1145. There's a known interpreter issue that is causing them to be exposed as properties even if their usage is set to `Excluded`.

There are some minimal changes to subgraph outputs as well. The subgraph output node previously used the same type list as the Blackboard, which now includes some types that the subgraph output node shouldn't have.

---
### Testing status

- Manual testing to make sure blackboard properties can still be added and used. Formal TestRail/automated cases will be created after UI is in place.

### CODE REVIEW

**Review by brett-groshong — APPROVED** [review]

**Comment by esmelusina** [comment]
Interesting behavior, different inspected fields for the reference node and the selected blackboard item:
Node,
image
Variable,
image

I suspect we just want to forward the inspected fields from the variable _for_ the reference node, since we are highlighting the blackboard property whenever you are manipulating the reference node. Probably a design question though, if you want to poke Ben/Alex/Tracy on correct behavior here.

**Comment by esmelusina** [comment]
Looks like there is a regression on the test nodes for sampling textures and gradients-- I might look at fixing that regression here so the variables are fully testable.

Also-- it looks like the SampleTexture2D nodes that tech art authored either are regressed or maybe malformed-- I think those can be addressed in a separate PR though.

File: com.unity.sg2/Editor/GraphDeltaRegistry/Registry/ContextBuilder.cs [hunk 1]
```diff
@@ -1,18 +1,33 @@
...
+
+            NodeBuilderUtils.ParameterDescriptorToField(
+                new ParameterDescriptor($"out_{fieldName}", type, GraphType.Usage.Out),
+                TYPE.Vec4,
+                contextNode,
+                registry);
+        }
 
-        static public void AddContextEntry(NodeHandler contextNode, ContextEntry entry, Registry registry)
+        public static void AddContextEntry(NodeHandler contextNode, ContextEntry entry, Registry registry)
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphDeltaRegistry/Registry/ContextBuilder.cs:30]: May as well mark this as obsolete.

File: com.unity.sg2/Editor/GraphDeltaRegistry/Registry/ContextBuilder.cs [hunk 2]
```diff
@@ -25,23 +40,43 @@ static public void AddContextEntry(NodeHandler contextNode, ContextEntry entry,
...
-            if(displayName != null)
+            if (displayName != null)
                 port.AddField(kDisplayName, displayName);
-            if(defaultValue != null)
+            if (defaultValue != null)
                 port.AddField(kDefaultValue, defaultValue);
+        }
 
+        public static void AddReferableEntry(NodeHandler contextNode,
+            ContextEntry entry,
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphDeltaRegistry/Registry/ContextBuilder.cs:58]: Same here, can probably mark this one obsolete to.

File: com.unity.sg2/Editor/GraphDeltaRegistry/Registry/ContextBuilder.cs [hunk 3]
```diff
@@ -1,18 +1,33 @@
...
 
         public RegistryFlags GetRegistryFlags() => RegistryFlags.Base;
 
+        public static void AddContextEntry(NodeHandler contextNode, ITypeDescriptor type, string fieldName, Registry registry)
+        {
+            NodeBuilderUtils.ParameterDescriptorToField(
+                new ParameterDescriptor(fieldName, type, GraphType.Usage.In),
+                TYPE.Vec4,
+                contextNode,
+                registry);
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphDeltaRegistry/Registry/ContextBuilder.cs:21]: `ParameterDescriptorToField` returns the `PortHandler` that is created. You could pass return that from `AddContextEntry` and pass it directly into the SetupReferableEntry.

**Comment by esmelusina** [comment]
Added the test nodes for textures and gradients back in; changing the texture value on a property doesn't propagate from the context output port at the moment.
image

**Comment by esmelusina** [comment]
Looks like there is an issue in the interpreter with unexposed & constant referables. I'll have to talk to @elizabeth-legros to catch up with what's going on here, but the value daisy chaining is ignoring the material property block's output in this case. I suspect it's just because one uses the full path name and one just uses the port name, but I'm not sure.

```
struct ShaderGraphBlock_MaterialPropertyContextOutput
{
    Gradient _d7ff764b4e5826ea9785ccf127464bd9; // uses port name -- unexposed & constant material property (eg. it's fake).
};

// Initializes correctly.
ShaderGraphBlock_MaterialPropertyContextBlock::ShaderGraphBlock_MaterialPropertyContextOutput SYNTAX_MaterialPropertyContextMain(ShaderGraphBlock_MaterialPropertyContextBlock::ShaderGraphBlock_MaterialPropertyContextInput In)
{
    ShaderGraphBlock_MaterialPropertyContextBlock::ShaderGraphBlock_MaterialPropertyContextOutput output;
    output._d7ff764b4e5826ea9785ccf127464bd9 = NewGradient(2, 8, 8, float4(0,0,0,0), float4(0,0,0,0), float4(0,0,0,0), float4(0,0,0,0), float4(0,0,0,0), float4(0,0,0,0), float4(0,0,0,0), float4(0,0,0,0), float2(0,0), float2(0,0), float2(0,0), float2(0,0), float2(0,0), float2(0,0), float2(0,0), float2(0,0));
    return output;
}

ShaderGraphBlock_MaterialPropertyContextBlock::ShaderGraphBlock_MaterialPropertyContextInput shadergraphblock_materialpropertycontextinput;
ShaderGraphBlock_MaterialPropertyContextBlock::ShaderGraphBlock_MaterialPropertyContextOutput shadergraphblock_materialpropertycontextoutput = SYNTAX_MaterialPropertyContextMain(shadergraphblock_materialpropertycontextinput);



struct SurfaceDescriptionCPInput
{
    float3 TangentSpaceNormal;
    Gradient ShaderGraphBlock_DefaultContextDescriptor__d7ff764b4e5826ea9785ccf127464bd9; // uses long name
};

ShaderGraphBlock_DefaultContextDescriptorBlock::ShaderGraphBlock_DefaultContextDescriptorInput shadergraphblock_defaultcontextdescriptorinput;

// value daisy chains from surface description, but not from the material property block
shadergraphblock_defaultcontextdescriptorinput._d7ff764b4e5826ea9785ccf127464bd9 = surfacedescriptioncpinput.ShaderGraphBlock_DefaultContextDescriptor__d7ff764b4e5826ea9785ccf127464bd9;
ShaderGraphBlock_DefaultContextDescriptorBlock::ShaderGraphBlock_DefaultContextDescriptorOutput shadergraphblock_defaultcontextdescriptoroutput = SYNTAX_DefaultContextDescriptorMain(shadergraphblock_defaultcontextdescriptorinput);
```

**Review by esmelusina — APPROVED** [review]

**Pull request merged**